File: apps/betty-server/src/trips/application/use-cases/create-trip/create-trip.use-case.ts
import { Injectable, Inject } from '@nestjs/common';
import { TripRepository } from '../../../domain/repositories/trip.repository';
import { Trip } from '../../../domain/entities/trip.entity';
import { TRIP_REPOSITORY } from '../../../infrastructure/ioc/symbols';
import { randomUUID } from 'crypto';

@Injectable()
export class CreateTripUseCase {
  constructor(
    @Inject(TRIP_REPOSITORY)
    private readonly tripRepository: TripRepository
  ) {}

  async execute(name: string): Promise<Trip> {
    // Verificar que no haya un viaje en progreso
    const hasInProgress = await this.tripRepository.hasInProgress();
    if (hasInProgress) {
      throw new Error('There is already a trip in progress');
    }

    // Crear nuevo viaje
    const trip = new Trip(
      randomUUID(),
      name,
      new Date() // startedAt
    );

    return this.tripRepository.create(trip);
  }
}

File: apps/betty_app/lib/trips/application/use_cases/create_trip_use_case.dart
import '../../infrastructure/services/trip_api_service.dart';
import '../../domain/entities/trip.dart';

class CreateTripUseCase {
  final TripApiService _apiService;

  const CreateTripUseCase(this._apiService);

  Future<Trip> execute(String name) async {
    return await _apiService.createTrip(name);
  }
}
